REMEMBER: FIRST CREATE A COPY OF THIS FILE WITH A UNIQUE NAME AND DO YOUR WORK THERE. AND MAKE SURE YOU COMMIT YOUR CHANGES TO THE `hw3_submissions` BRANCH.

# Assignment 3 | Cleaning and Exploring Data with Pandas



<img src="data/scoreCard.jpg" width=250>

In this assignment, you will investigate restaurant food safety scores for restaurants in San Francisco. Above is a sample score card for a restaurant. The scores and violation information have been made available by the San Francisco Department of Public Health. 

## Loading Food Safety Data


There are 2 files in the data directory:
1. business.csv containing food establishments in San Francisco
1. inspections.csv containing retaurant inspections records

Let's start by loading them into Pandas dataframes.  One of the files, business.csv, has encoding (ISO-8859-1), so you will need to account for that when reading it.

### Question 1

#### Question 1a
Read the two files noted above into two pandas dataframes named `bus` and `ins`, respectively. Print the first 5 rows of each to inspect them.


In [1]:
import pandas as pd

In [2]:
bus = pd.read_csv('data/businesses.csv', encoding='ISO-8859-1')
ins = pd.read_csv('data/inspections.csv')

In [3]:
bus.head()

business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           31      NORMAN'S ICE CREAM AND FREEZES   
4           45                 CHARLIE'S DELI CAFE   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   
4                3202 FOLSOM ST   San Francisco    CA       94110  37.747114   

    longitude  phone_number  
0 -122.403816  +14154217044  
1 -122.421547  +14157763262  
2 -122.403135  +14156779494  
3 -122.419004           NaN  
4 -122.413641  +14156415051

In [4]:
ins.head()

business_id  score      date     type
0           10     82  20160503  routine
1           10     94  20140729  routine
2           10     92  20140114  routine
3           19     94  20160513  routine
4           19     94  20141110  routine

## Examining the Business data

From its name alone, we expect the `businesses.csv` file to contain information about the restaurants. Let's investigate this dataset.

### Question 2

#### Question 2a: How many records are there?

In [5]:
rec_num = bus.count()

print(rec_num)

business_id     6315
name            6315
address         6315
city            6315
state           6315
postal_code     6056
latitude        3599
longitude       3599
phone_number    5696
dtype: int64


#### Question 2b: How many unique business IDs are there?  

In [39]:
uni_cnt = bus['business_id'].unique()

print(len(bus),len(uni_cnt),"\nAre they equal? =",len(ins)==len(uni_cnt))

6315 6315 
Are they equal? = False


#### Question 2c: What are the 5 most common businesses by name, and how many are there in San Francisco?

In [7]:
cnt_names = bus['name'].value_counts()

cnt_names.head()

STARBUCKS COFFEE              72
PEET'S COFFEE & TEA           24
MCDONALDS                     12
WALGREENS                     11
SAN FRANCISCO SOUP COMPANY    11
Name: name, dtype: int64

## Zip code

Next, let's  explore some of the variables in the business table. We begin by examining the postal code.

### Question 3

#### Question 3a
How are the zip code values stored in python (i.e. data type)?

To answer this you might want to examine a particular entry.

In [8]:
entry_1 = bus.iloc[1,:]

type (entry_1.loc['postal_code'])

str

#### Question 3b

What are the unique values of postal_code?

In [9]:
postal_codes = bus['postal_code'].unique()

print(postal_codes)

['94104' '94109' '94133' '94110' '94122' '94103' '94115' '94131' '94111'
 '94117' '94107' '94108' '94102' '94132' '94105' '94134' '94124' '94116'
 '94121' '94118' '94112' '94123' '94114' '94127' '941102019' '941' '94545'
 '94066' '941033148' nan '94158' '95105' '94013' '94130' 'CA' '92672'
 '94120' '94143' '94609' '94101' '00000' '94188' '94621' '94014' '94129'
 '94602' 'Ca']


#### Question 3c

Let's say we decide to exclude the businesses that have no zipcode for our analysis (which might include food trucks for example).  Use the list of valid 5-digit zip codes below to create a new dataframe called bus_valid, with only businesses whose postal_codes show up in this list of valid zipcodes. How many businesses are there in this new dataframe?

In [10]:
validZip = ["94102", "94103", "94104", "94105", "94107", "94108",
            "94109", "94110", "94111", "94112", "94114", "94115",
            "94116", "94117", "94118", "94121", "94122", "94123", 
            "94124", "94127", "94131", "94132", "94133", "94134"]

In [11]:
vb = bus[bus['postal_code'].isin(validZip)]

## Latitude and Longitude

Another aspect of the data we want to consider is the prevalence of missing values. If many records have missing values then we might be concerned about whether the nonmissing values are representative of the population.

### Question 4
 
Consider the longitude and latitude in the business DataFrame. 

#### Question 4a

How many businesses are missing longitude values, working with only the businesses that are in the list of valid zipcodes?

In [12]:
vb[pd.isnull(vb['longitude'])]

business_id                               name             address  \
200           821                       LA SANTANECA    3781 MISSION ST    
210           852  NEW HOLLYWOOD BAKERY & RESTAURANT    652 PACIFIC AVE    
760          2934                   LOU'S FISH SHACK   300 JEFFERSON ST    
826          3123                        LES JOULINS        44 ELLIS ST    
1100         4323                             DIVA'S       1081 POST ST    
...           ...                                ...                 ...   
6310        90234               ACCESS SFUSD THE ARC     1520 HOWARD ST    
6311        90236                     PHO DONG HUONG  667 MONTEREY BLVD    
6312        90237                   WIN'S RESTAURANT    3040 TARAVAL ST    
6313        90238                   USA POWER MARKET     1524 OCEAN AVE    
6314        90239                         ALBA RAY'S    2293 MISSION ST    

               city state postal_code  latitude  longitude  phone_number  
200   San Francisco    CA       94110       NaN        NaN  +14156481034  
210   San Francisco    CA       94133       NaN        NaN  +14153979919  
760   San Francisco    CA       94133       NaN        NaN  +14157715687  
826   San Francisco    CA       94102       NaN        NaN  +14153975397  
1100  San Francisco    CA       94109       NaN        NaN  +14154743482  
...             ...   ...         ...       ...        ...           ...  
6310  San Francisco    CA       94103       NaN        NaN           NaN  
6311  San Francisco    CA       94127       NaN        NaN  +14154945048  
6312  San Francisco    CA       94116       NaN        NaN  +14158606036  
6313  San Francisco    CA       94112       NaN        NaN  +14153331878  
6314  San Francisco    CA       94110       NaN        NaN  +14158881393  

[2483 rows x 9 columns]

#### Question 4b

Create a new dataframe with one row for each valid zipcode. The dataframe should include the following three columns:

1. `postal_code`: Contains the zip codes in the `validZip` variable above.
2. `null_lon`: The number of businesses in that zipcode with missing `longitude` values.
3. `not_null_lon`: The number of businesses without missing `longitude` values.

In [13]:
null_lon_df = vb[pd.isna(vb['longitude'])]
not_null_lon_df = vb[pd.notna(vb['longitude'])]

In [14]:
print(null_lon_df)

      business_id                               name             address  \
200           821                       LA SANTANECA    3781 MISSION ST    
210           852  NEW HOLLYWOOD BAKERY & RESTAURANT    652 PACIFIC AVE    
760          2934                   LOU'S FISH SHACK   300 JEFFERSON ST    
826          3123                        LES JOULINS        44 ELLIS ST    
1100         4323                             DIVA'S       1081 POST ST    
...           ...                                ...                 ...   
6310        90234               ACCESS SFUSD THE ARC     1520 HOWARD ST    
6311        90236                     PHO DONG HUONG  667 MONTEREY BLVD    
6312        90237                   WIN'S RESTAURANT    3040 TARAVAL ST    
6313        90238                   USA POWER MARKET     1524 OCEAN AVE    
6314        90239                         ALBA RAY'S    2293 MISSION ST    

               city state postal_code  latitude  longitude  phone_number  
200   San Fr

In [15]:
null_lon_gb=null_lon_df.groupby('postal_code').size().sort_values(ascending=False).reset_index(name='null_lon')

In [16]:
print(null_lon_gb)

   postal_code  null_lon
0        94103       252
1        94110       251
2        94107       247
3        94102       196
4        94109       146
5        94133       135
6        94122       123
7        94111       115
8        94114       108
9        94105       100
10       94118       100
11       94124        88
12       94108        82
13       94115        76
14       94117        67
15       94132        65
16       94112        59
17       94104        59
18       94121        55
19       94123        53
20       94116        35
21       94134        29
22       94127        28
23       94131        14


In [17]:
not_null_lon_gb=not_null_lon_df.groupby('postal_code').size().sort_values(ascending=False).reset_index(name='not_null_lon')

In [18]:
print(not_null_lon_gb)

   postal_code  not_null_lon
0        94110           334
1        94103           306
2        94133           282
3        94102           262
4        94109           237
5        94107           204
6        94111           167
7        94115           162
8        94122           147
9        94118           143
10       94108           142
11       94112           131
12       94117           130
13       94105           126
14       94114           122
15       94123           116
16       94121           106
17       94124            76
18       94104            74
19       94132            69
20       94116            58
21       94134            45
22       94127            43
23       94131            34


In [19]:
null_lon_gb['postal_code']=null_lon_gb['postal_code'].astype(str)
not_null_lon_gb['postal_code']=not_null_lon_gb['postal_code'].astype(str)

In [20]:
zip_lon = null_lon_gb.merge(not_null_lon_gb, how='inner',
                          on='postal_code')

In [21]:
print(zip_lon)

   postal_code  null_lon  not_null_lon
0        94103       252           306
1        94110       251           334
2        94107       247           204
3        94102       196           262
4        94109       146           237
5        94133       135           282
6        94122       123           147
7        94111       115           167
8        94114       108           122
9        94105       100           126
10       94118       100           143
11       94124        88            76
12       94108        82           142
13       94115        76           162
14       94117        67           130
15       94132        65            69
16       94112        59           131
17       94104        59            74
18       94121        55           106
19       94123        53           116
20       94116        35            58
21       94134        29            45
22       94127        28            43
23       94131        14            34


#### 4c. Do any zip codes appear to have more than their 'fair share' of missing longitude? 

To answer this, you will want to compute the proportion of missing longitude values for each zip code, and print the proportion missing longitude, and print the top five zipcodes in descending order of proportion missing postal_code.


In [3]:
zip_lon.insert(3,'prop_miss',(zip_lon['null_lon']/(zip_lon['null_lon']+zip_lon['not_null_lon'])))

NameError: name 'zip_lon' is not defined

In [31]:
zip_lon_order = zip_lon.sort_values(by='prop_miss',ascending=False)
print(zip_lon_order)

   postal_code  null_lon  not_null_lon  prop_miss
2        94107       247           204   0.547672
11       94124        88            76   0.536585
15       94132        65            69   0.485075
8        94114       108           122   0.469565
6        94122       123           147   0.455556
0        94103       252           306   0.451613
17       94104        59            74   0.443609
9        94105       100           126   0.442478
1        94110       251           334   0.429060
3        94102       196           262   0.427948
10       94118       100           143   0.411523
7        94111       115           167   0.407801
22       94127        28            43   0.394366
21       94134        29            45   0.391892
4        94109       146           237   0.381201
20       94116        35            58   0.376344
12       94108        82           142   0.366071
18       94121        55           106   0.341615
14       94117        67           130   0.340102


# Investigate the inspection data

Let's now turn to the inspection DataFrame. Earlier, we found that `ins` has 4 columns, these are named `business_id`, `score`, `date` and `type`.  In this section, we determine the granularity of `ins` and investigate the kinds of information provided for the inspections. 

### Question 5

#### Question 5a
As with the business data, assess whether there is one inspection record for each business, by counting how many rows are in the data and how many unique businesses there are in the data. If they are exactly the same number, it means there is only one inspection per business, clearly.

In [4]:
unq_ins = ins['business_id'].unique()

print("Total Inspections=",len(ins),"Total uniqe insp=",len(unq_ins),"\nAre they equal? =",len(ins)==len(unq_ins))

Total Inspections= 15430 Total uniqe insp= 5730 
Are they equal? = False


#### Question 5b

What values does `type` take on? How many occurrences of each value is in the DataFrame? Create a new dataframe named `ins2` by copying `ins` and keeping only records with values of `type` that occur more than 10 times in the original table. In other words, eliminate records that have values of `type` that occur rarely (< 10 times). Check the result to make sure rare types are eliminated.

In [3]:
ins_type = ins['type'].value_counts()

print(ins_type)

routine      15429
complaint        1
Name: type, dtype: int64


In [20]:
ins2 = ins2.groupby('type').filter(lambda x: len(x) > 10)

print(ins2['type'].value_counts())

routine    15429
Name: type, dtype: int64


#### Question 5c

Since the data was stored in a .csv file, the dates are formatted as strings such as `20160503`. Once we read in the data, we would like to have dates in an appropriate format for analysis. Add a new column called `year` by capturing the first four characters of the date column. 

Hint: we have seen multiple ways of doing this in class, includings `str` operations, `lambda` functions, `datetime` operations, and others. Choose the method that works best for you :)

In [5]:
from datetime import datetime

In [6]:
ins2.insert(4,'year',pd.to_datetime(ins2['date'], format='%Y%m%d').dt.strftime('%Y'))
#turn date column into a date time
#add to ins2 as a new column 'year'

In [12]:
ins2.head()

business_id  score      date     type  year
0           10     82  20160503  routine  2016
1           10     94  20140729  routine  2014
2           10     92  20140114  routine  2014
3           19     94  20160513  routine  2016
4           19     94  20141110  routine  2014

#### Question 5d

What range of years is covered in this data set? Are there roughly same number of inspections each year? Try dropping records for any years with less than 50 inspections and store the result in a new dataframe named `ins3`.

In [9]:
ins2.year.value_counts()

2016    5839
2014    5629
2015    3923
2013      38
Name: year, dtype: int64

In [19]:
ins3 = ins2.groupby('year').filter(lambda x: len(x) > 50)

ins3.year.value_counts()

2016    5839
2014    5629
2015    3923
Name: year, dtype: int64

Let's examine only the inspections for one year: 2016. This puts businesses on a more equal footing because [inspection guidelines](https://www.sfdph.org/dph/eh/Food/Inspections.asp) generally refer to how many inspections should occur in a given year.

In [23]:
ins_2016 = ins3.loc[ins3['year']=='2016'] #creates new df that only holds the rows in ins3 that have 2016 as the year

ins_2016.head()

business_id  score      date     type  year
0            10     82  20160503  routine  2016
3            19     94  20160513  routine  2016
6            24     98  20161005  routine  2016
7            24     96  20160311  routine  2016
10           31     84  20160816  routine  2016

### Question 6

#### Question 6a

Merge the business and 2016 inspections data, keeping all businesses regardless of whether they show up in the inspections file. Show the first several rows of the resulting dataframe.

In [29]:
bus_ins = bus.merge(ins_2016, how='left',
                          on='business_id')

print(bus_ins[:20])

    business_id                                name  \
0            10                    TIRAMISU KITCHEN   
1            19               NRGIZE LIFESTYLE CAFE   
2            24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3            24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
4            31      NORMAN'S ICE CREAM AND FREEZES   
5            45                 CHARLIE'S DELI CAFE   
6            45                 CHARLIE'S DELI CAFE   
7            48                          ART'S CAFE   
8            50                          SUSHI ZONE   
9            54                 RHODA GOLDMAN PLAZA   
10           56                          CAFE X + O   
11           58                         OASIS GRILL   
12           61                            CHOWDERS   
13           66                    STARBUCKS COFFEE   
14           66                    STARBUCKS COFFEE   
15           67                     REVOLUTION CAFE   
16           67                     REVOLUTION CAFE   
17        

#### Question 6b
Print the 20 lowest rated businesses names, their addresses, and their ratings.

In [48]:
bus_lowest_rate = bus_ins.sort_values(by='score',ascending=True)
bus_lowest_rate

business_id                     name             address           city  \
4892        71373  GOLDEN RIVER RESTAURANT    5827 GEARY BLVD   San Francisco   
7324        87652                POKI TIME    2101 LOMBARD ST   San Francisco   
6281        81667             CRAZY PEPPER  2257 SAN JOSE AVE   San Francisco   
4873        71232               UNCLE CAFE      65 WAVERLY PL   San Francisco   
3008        34585         HING WANG BAKERY       339 JUDAH ST   San Francisco   
...           ...                      ...                 ...            ...   
7606        90234     ACCESS SFUSD THE ARC     1520 HOWARD ST   San Francisco   
7607        90236           PHO DONG HUONG  667 MONTEREY BLVD   San Francisco   
7608        90237         WIN'S RESTAURANT    3040 TARAVAL ST   San Francisco   
7609        90238         USA POWER MARKET     1524 OCEAN AVE   San Francisco   
7610        90239               ALBA RAY'S    2293 MISSION ST   San Francisco   

     state postal_code   latitude   longitude  phone_number  score  \
4892    CA       94121  37.780144 -122.482040  +14156685105   52.0   
7324    CA       94123        NaN         NaN  +14156366868   55.0   
6281    CA       94112        NaN         NaN  +14157348687   55.0   
4873    CA       94108  37.794034 -122.406794  +14155701380   55.0   
3008    CA       94122  37.762174 -122.465809  +14156813928   56.0   
...    ...         ...        ...         ...           ...    ...   
7606    CA       94103        NaN         NaN           NaN    NaN   
7607    CA       94127        NaN         NaN  +14154945048    NaN   
7608    CA       94116        NaN         NaN  +14158606036    NaN   
7609    CA       94112        NaN         NaN  +14153331878    NaN   
7610    CA       94110        NaN         NaN  +14158881393    NaN   

            date     type  year  
4892  20161031.0  routine  2016  
7324  20160721.0  routine  2016  
6281  20160901.0  routine  2016  
4873  20160308.0  routine  2016  
3008  20160201.0  routine  2016  
...          ...      ...   ...  
7606         NaN      NaN   NaN  
7607         NaN      NaN   NaN  
7608         NaN      NaN   NaN  
7609         NaN      NaN   NaN  
7610         NaN      NaN   NaN  

[7611 rows x 13 columns]

In [49]:
bus_low20 = bus_lowest_rate[:20][["name","address","score"]]

bus_low20

name              address  score
4892      GOLDEN RIVER RESTAURANT     5827 GEARY BLVD    52.0
7324                    POKI TIME     2101 LOMBARD ST    55.0
6281                 CRAZY PEPPER   2257 SAN JOSE AVE    55.0
4873                   UNCLE CAFE       65 WAVERLY PL    55.0
3008             HING WANG BAKERY        339 JUDAH ST    56.0
5627                   GOLDEN WOK   295 B ORIZABA AVE    56.0
5383                   TOP'S CAFE  2460 SAN BRUNO AVE    57.0
6067    L & G VIETNAMESE SANDWICH         602 EDDY ST    57.0
836                  RED A BAKERY      634 CLEMENT ST    57.0
4209      SAN TUNG RESTAURANT LLC      1031 IRVING ST    57.0
549           MA'S DIM SUM & CAFE      1315 POWELL ST    57.0
4903  NEW GARDEN RESTAURANT, INC.       716 KEARNY ST    57.0
6645    KOBANI MEDITERANEAN GRILL       3242 SCOTT ST    58.0
624           JOANIE'S DINER INC.   1329 COLUMBUS AVE    59.0
6938         PAK NAM THAI CUISINE       655 LARKIN ST    59.0
3563                  JADE GARDEN      3608 BALBOA ST    59.0
4234                        ITHAI         720 POST ST    59.0
6151          VALLARTA'S TACO BAR     2022 MISSION ST    60.0
4850             HONG KONG LOUNGE     5322 GEARY BLVD    60.0
5013                 WOLFES LUNCH        1220 16TH ST    60.0

## Done!

Now commit this notebook to your `hw3_submissions` branch, push it to your GitHub repo, and open a PR!